coa/pkg/apis/v1alpha2/providers/stringlock/memorystringlock/memorystringlock.go

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ * SPDX-License-Identifier: MIT
+ */
+
+package memorystringlock
+
+import (
+	"encoding/json"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/eclipse-symphony/symphony/coa/pkg/apis/v1alpha2/providers"
+	"github.com/eclipse-symphony/symphony/coa/pkg/apis/v1alpha2/providers/stringlock"
+	"github.com/eclipse-symphony/symphony/coa/pkg/logger"
+)
+
+var sLog = logger.NewLogger("coa.runtime")
+
+type MemoryStringLockProvider struct {
+	lm            *LockManager
+	cleanInterval int //seconds
+	purgeDuration int // 12 hours before
+}
+
+type MemoryStringLockProviderConfig struct {
+	CleanInterval int `json:"cleanInterval"`
+	PurgeDuration int `json:"purgeDuration"`
+}
+
+func toMemoryStringLockProviderConfig(config providers.IProviderConfig) (MemoryStringLockProviderConfig, error) {
+	ret := MemoryStringLockProviderConfig{}
+	data, err := json.Marshal(config)
+	if err != nil {
+		return ret, err
+	}
+	err = json.Unmarshal(data, &ret)
+	return ret, err
+}
+
+func (mslp *MemoryStringLockProvider) Init(config providers.IProviderConfig) error {
+	stringLockConfig, err := toMemoryStringLockProviderConfig(config)
+	if err != nil {
+		sLog.Errorf("  P (String Lock): failed to parse provider config %+v", err)
+		return errors.New("expected MemoryStringLockProviderConfig")
+	}
+	if stringLockConfig.CleanInterval > 0 {
+		mslp.cleanInterval = stringLockConfig.CleanInterval
+	} else {
+		mslp.cleanInterval = 30 // default: 30 seconds
+	}
+	if stringLockConfig.PurgeDuration > 0 {
+		mslp.purgeDuration = stringLockConfig.PurgeDuration
+	} else {
+		mslp.purgeDuration = 60 * 60 * 12 // default: 12 hours
+	}
+	mslp.lm = NewLockManager()
+	go func() {
+		for {
+			mslp.lm.clean(-mslp.purgeDuration)
+			time.Sleep(time.Duration(mslp.cleanInterval) * time.Second)
+		}
+	}()
+	return nil
+}
+
+func (mslp *MemoryStringLockProvider) Lock(key string) stringlock.UnLock {
+	mslp.lm.getLockNode(key).Lock()
+
+	return func() {
+		mslp.lm.getLockNode(key).Unlock()
+		go mslp.lm.updateLockLRU(key)
+	}
+}
+
+type LockNode struct {
+	lastUsedTime time.Time
+	key          string
+	lock         sync.Mutex
```

Comment:
Why maintain the lock as linked list? Seems GET lock is not efficient.

Reply:
I share the same concerns with Lency. I can understand you want to keep LRU order for purging. How about we keep a map for efficient GET?

Reply:
getlock will using a concurrent map 

```
locknode, _ := lm.m.LoadOrStore(key, &LockNode{key: key, prev: nil, next: nil})
```